Privacy Choices: Handle Save CTA

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8948
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds support for taping the `Save` button on the Privacy Banner.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Hint: you can use the new option in `Developer Options` to trigger showing the Privacy Banner again. See here: p1684926805528519-slack-C02KUCFCSFP

#### TC1 WPCOM account

1. Log in as WPCOM user
2. On the Privacy Banner, change the "Analytics" switch
3. Tap "Save"
4. Assert that after a second, the Privacy Banner is dismissed
5. Go to https://wordpress.com/me/privacy and check if the Privacy Settings were updated
6. Restart the app
7. Assert that Privacy Banner is not shown again

#### TC2 Non-WPCOM account

1. Log in as non-WPCOM user
2. On the Privacy Banner, change the "Analytics" switch
3. Tap "Save"
4. Assert that the Privacy Banner is dismissed
5. Assert that Tracking is disabled (no "Tracked :" logs)
6. Restart the app
7. Assert that the Privacy Banner is not shown again

#### TC3 WPCOM account but error

1. Log in as WPCOM user
2. Turn off the internet connection
3. On the Privacy Banner, change the "Analytics" switch
4. Tap "Save"
5. Assert that the Privacy Banner is dismissed and Snackbar with "Retry" button is shown
6. Turn on the internet connection
7. Tap on "Retry"
8. Assert that Snack bar is hidden
9. Go to https://wordpress.com/me/privacy and check if the Privacy Settings were updated


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

#### Success

https://github.com/woocommerce/woocommerce-android/assets/5845095/a1877fc4-ea89-4184-9126-240bd7c2b9d9

#### Failure (and then success)

https://github.com/woocommerce/woocommerce-android/assets/5845095/e441632c-4818-4a11-bfcb-70e918ef1814



- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
feat: save user choice after pressing Save CTA

wzieba committed:
feat: add saved privacy banner settings preference toggle to dev options

wzieba committed:
feat: handle retry action

wzieba committed:
Merge branch 'trunk' into issue/8948_handle_save_cta

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt:
```diff
@@ -251,6 +253,18 @@ class MainActivityViewModel @Inject constructor(
         )
     }
 
+    fun onPrivacyPreferenceUpdateFailed(analyticsEnabled: Boolean) {
+        triggerEvent(ShowPrivacyPreferenceUpdatedFailed(analyticsEnabled))
+    }
```
np: As this function only performs one action and can be called only from the `onRequestPrivacyUpdate`, WDYT about merging the two functions? :

```
fun onRequestPrivacyUpdate(analyticsEnabled: Boolean) {
        launch {
            privacyRepository.updateTracksSetting(analyticsEnabled).onFailure {
                triggerEvent(ShowPrivacyPreferenceUpdatedFailed(analyticsEnabled))
            }
        }
    }
```

wzieba replied:
Good point, thanks!

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/privacy/banner/PrivacyBannerViewModel.kt:
```diff
@@ -1,26 +1,70 @@
+    fun onSavePressed() {
+        val analyticsPreference = _state.value?.analyticsSwitchEnabled ?: false
+
+        launch {
+            if (repository.isUserWPCOM()) {
+                _state.value = _state.value?.copy(loading = true)
+                val event =
+                    repository.updateTracksSetting(_state.value?.analyticsSwitchEnabled ?: false)
+                _state.value = _state.value?.copy(loading = false)
+
+                event.fold(
+                    onSuccess = {
+                        appPrefsWrapper.savedPrivacyBannerSettings = true
+                        analyticsTrackerWrapper.sendUsageStats = analyticsPreference
+                        triggerEvent(Dismiss)
+                    },
+                    onFailure = {
+                        triggerEvent(ShowError(requestedChange = analyticsPreference))
+                    }
+                )
+            } else {
+                appPrefsWrapper.savedPrivacyBannerSettings = true
+                analyticsTrackerWrapper.sendUsageStats = analyticsPreference
+            }
```
How do you feel about unit testing this logic?

wzieba replied:
I agree it'd be valuable, but at this moment, if you don't mind, I'd prefer to merge this PR without unit tests, add the missing feature and then take care of tests. I know it's not ideal, but I'd love to merge things before tomorrow.

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/DeveloperOptionsViewModel.kt:
```diff
@@ -25,14 +26,25 @@ class DeveloperOptionsViewModel @Inject constructor(
+    private val appPrefsWrapper: AppPrefsWrapper,
 ) : ScopedViewModel(savedState) {
+
+    private val savedPrivacySettingsOnDialogItem = ToggleableListItem(
```
Nice! 🙏

wzieba committed:
refactor: simplify data flow in Main VM

wzieba commented:
Thanks for the review @atorresveiga ! I've applied your suggestion here fcd271296630001a7f6341525ae21584315a0f0e , ready for the next round.

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/privacy/banner/PrivacyBannerViewModel.kt:
```diff
@@ -1,26 +1,70 @@
+                        triggerEvent(ShowError(requestedChange = analyticsPreference))
+                    }
+                )
+            } else {
```
Do you think we should trigger a Dismiss event when the user is not a WPCOM user? Otherwise the banner will not hide
`triggerEvent(Dismiss)`

wzieba replied:
Yes, you're completely right. Thanks! Fixed in e5917fd7c2fa5f7e435ee0cc99f7c3b18b90d1a8

atorresveiga commented:
np: When the user is a WPCOM user, and the save Settings fails, the app shows the retry SnackBar. If I turn the internet connection on again, press retry, and the retry succeeds, the setting is updated, but when I restart the app, the banner will show again. Shouldn't the banner be hidden by default after the retry? Here is a video:

https://github.com/woocommerce/woocommerce-android/assets/18119390/648d3fa7-8f7e-412d-b05a-177e44544ed3

atorresveiga commented:
One last thing that I'm not sure is related to this PR. When I logged in as a no-WPCOM user, I used the Aruba user (thanks for that 🙏 ) that is a no-WPCOM user but uses the email of a WPCOM user. The issue was that I could update the Analytic setting using the privacy banner but those changes were not reflected on the privacy settings screen (it seems that is always using the settings from the WPCOM user).

wzieba committed:
fix: dismiss dialog if non-wpcom user updated analytic preferences

wzieba committed:
Revert "refactor: simplify data flow in Main VM"

This reverts commit fcd271296630001a7f6341525ae21584315a0f0e.

wzieba committed:
fix: mark privacy banner as seen if user updated settings after retry

wzieba commented:
> Shouldn't the banner be hidden by default after the retry?

Yes it should, good point! Fixed in 4e9f350f55c6d341e4677b79a3adfbda19a59b8b

----

I've also reverted the previous suggestion (here da49164865a24610a8086f33c797eedff8ed1c95) because it changed the flow of events 

from

1. `PrivacyBannerViewModel` fails requests
2. `PrivacyBannerFragment` calls `mainViewModel#onPrivacyPreferenceUpdateFailed`
3. `mainViewModel` triggers showing `Snackbar`
4. The app waits for user action. If user presses "Retry" then we retry the request

to

1. `PrivacyBannerViewModel` fails requests
2. `PrivacyBannerFragment` calls `mainViewModel#onPrivacyPreferenceUpdateFailed`
3. `mainViewModel` triggers `privacyRepository#updateTracksSetting` again (the same what `PrivacyBannerViewModel` just did)
4. The app only then, if request fails, shows Retry snackbar

It's a minor difference if we test "no internet connection" case, but it might be an issue if someone has poor internet connection and there's some HTTP issue (like time-out).

---

> One last thing that I'm not sure is related to this PR.

Thank you for this report! Yes, you're right - this issue shouldn't be caused by this PR. I'll make some tests and create a separate issue.

atorresveiga reviewed (approved):
Everything is working as expected!
LGTM! :shipit:
